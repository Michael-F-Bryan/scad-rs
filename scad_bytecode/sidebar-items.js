window.SIDEBAR_ITEMS = {"constant":[["EXECUTABLE","The name of the executable called via a [`Program`]’s shebang line."],["VERSION","This crate’s version."]],"enum":[["Constant",""],["Instruction","A single instruction executed by the virtual machine."]],"struct":[["Chunk","A set of instructions."],["Disassembler","An indent-aware pretty-printer."],["Program",""]]};