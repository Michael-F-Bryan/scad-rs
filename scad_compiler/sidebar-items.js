window.SIDEBAR_ITEMS = {"enum":[["Severity","How severe a diagnostic is."]],"mod":[["lowering",""],["parsing",""]],"struct":[["Database",""],["Diagnostic","A message from the compiler."],["Diagnostics","A set of [`Diagnostic`]s."],["Location",""],["Spanned",""],["Text","A reference-counted string."]],"trait":[["IntoDiagnostic",""]]};