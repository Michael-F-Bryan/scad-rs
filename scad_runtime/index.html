<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The OpenSCAD virtual machine and runtime environment."><meta name="keywords" content="rust, rustlang, rust-lang, scad_runtime"><title>scad_runtime - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="../crates.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../scad_runtime/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../scad_runtime/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Crate scad_runtime</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.1.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul></div><section><div class="block"><ul><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></div></section></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../scad_runtime/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">scad_runtime</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/scad_runtime/lib.rs.html#1-21">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The OpenSCAD virtual machine and runtime environment.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.BuiltinFunction.html" title="scad_runtime::BuiltinFunction struct">BuiltinFunction</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ConversionError.html" title="scad_runtime::ConversionError struct">ConversionError</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Stack.html" title="scad_runtime::Stack struct">Stack</a></div><div class="item-right docblock-short"><p>The runtime stack.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.VirtualMachine.html" title="scad_runtime::VirtualMachine struct">VirtualMachine</a></div><div class="item-right docblock-short"><p>The OpenSCAD virtual machine.</p>
</div></div></div><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Geometry.html" title="scad_runtime::Geometry enum">Geometry</a></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.RuntimeError.html" title="scad_runtime::RuntimeError enum">RuntimeError</a></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Type.html" title="scad_runtime::Type enum">Type</a></div></div><div class="item-row"><div class="item-left module-item"><a class="enum" href="enum.Value.html" title="scad_runtime::Value enum">Value</a></div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Callbacks.html" title="scad_runtime::Callbacks trait">Callbacks</a></div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.Context.html" title="scad_runtime::Context trait">Context</a></div><div class="item-right docblock-short"><p>Contextual information provided to builtin functions.</p>
</div></div></div><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="fn" href="fn.prelude.html" title="scad_runtime::prelude fn">prelude</a></div><div class="item-right docblock-short"><p>All functions and constants that are available on startup.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="scad_runtime" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (b44197abb 2022-09-05)" ></div></body></html>