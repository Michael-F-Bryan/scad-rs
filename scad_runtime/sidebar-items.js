window.SIDEBAR_ITEMS = {"enum":[["Geometry",""],["RuntimeError",""],["Type",""],["Value",""]],"fn":[["prelude","All functions and constants that are available on startup."]],"struct":[["BuiltinFunction",""],["ConversionError",""],["Stack","The runtime stack."],["VirtualMachine","The OpenSCAD virtual machine."]],"trait":[["Callbacks",""],["Context","Contextual information provided to builtin functions."]]};