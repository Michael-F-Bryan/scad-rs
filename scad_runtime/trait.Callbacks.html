<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `Callbacks` trait in crate `scad_runtime`."><meta name="keywords" content="rust, rustlang, rust-lang, Callbacks"><title>Callbacks in scad_runtime - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../scad_runtime/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../scad_runtime/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Callbacks</a></h2><div class="sidebar-elems"><section><div class="block"><h3 class="sidebar-title"><a href="#provided-methods">Provided Methods</a></h3><ul><li><a href="#method.before_execute">before_execute</a></li><li><a href="#method.consume_geometry">consume_geometry</a></li><li><a href="#method.print">print</a></li></ul></div><div class="block"><h3 class="sidebar-title"><a href="#foreign-impls">Implementations on Foreign Types</a></h3><ul><li><a href="#impl-Callbacks-for-%26mut%20C">&amp;mut C</a></li><li><a href="#impl-Callbacks-for-Box%3CC%3E">Box&lt;C&gt;</a></li></ul></div><div class="block"><h3 class="sidebar-title"><a href="#implementors">Implementors</a></h3></div></section><h2 class="location"><a href="index.html">In scad_runtime</a></h2></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../scad_runtime/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></div></form></nav></div><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn"><span class="in-band">Trait <a href="index.html">scad_runtime</a>::<wbr><a class="trait" href="#">Callbacks</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/scad_runtime/callbacks.rs.html#5-28">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="docblock item-decl"><pre class="rust trait"><code>pub trait Callbacks {
    fn <a href="#method.before_execute" class="fnname">before_execute</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&amp;mut self,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_current_chunk: &amp;<a class="struct" href="../scad_bytecode/struct.Chunk.html" title="struct scad_bytecode::Chunk">Chunk</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_instruction_pointer: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.usize.html">usize</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_instruction: <a class="enum" href="../scad_bytecode/enum.Instruction.html" title="enum scad_bytecode::Instruction">Instruction</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;_stack: &amp;<a class="struct" href="struct.Stack.html" title="struct scad_runtime::Stack">Stack</a><br>&nbsp;&nbsp;&nbsp;&nbsp;) { ... }
<span class="item-spacer"></span>    fn <a href="#method.consume_geometry" class="fnname">consume_geometry</a>(&amp;mut self, _geometry: <a class="enum" href="enum.Geometry.html" title="enum scad_runtime::Geometry">Geometry</a>) { ... }
<span class="item-spacer"></span>    fn <a href="#method.print" class="fnname">print</a>(&amp;mut self, _values: &amp;[<a class="enum" href="enum.Value.html" title="enum scad_runtime::Value">Value</a>]) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.unit.html">()</a>, <a class="enum" href="enum.RuntimeError.html" title="enum scad_runtime::RuntimeError">RuntimeError</a>&gt; { ... }
}</code></pre></div><h2 id="provided-methods" class="small-section-header">Provided Methods<a href="#provided-methods" class="anchor"></a></h2><div class="methods"><details class="rustdoc-toggle" open><summary><div id="method.before_execute" class="method has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#12-19">source</a><h4 class="code-header">fn <a href="#method.before_execute" class="fnname">before_execute</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;&amp;mut self,<br>&nbsp;&nbsp;&nbsp;&nbsp;_current_chunk: &amp;<a class="struct" href="../scad_bytecode/struct.Chunk.html" title="struct scad_bytecode::Chunk">Chunk</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;_instruction_pointer: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.usize.html">usize</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;_instruction: <a class="enum" href="../scad_bytecode/enum.Instruction.html" title="enum scad_bytecode::Instruction">Instruction</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;_stack: &amp;<a class="struct" href="struct.Stack.html" title="struct scad_runtime::Stack">Stack</a><br>)</h4></div></summary><div class="docblock"><p>A callback that is fired immediately <em>before</em> executing the next
instruction.</p>
<p>At this point, the instruction pointer is still pointing at the current
instruction.</p>
</div></details><div id="method.consume_geometry" class="method has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#22">source</a><h4 class="code-header">fn <a href="#method.consume_geometry" class="fnname">consume_geometry</a>(&amp;mut self, _geometry: <a class="enum" href="enum.Geometry.html" title="enum scad_runtime::Geometry">Geometry</a>)</h4></div><div id="method.print" class="method has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#25-27">source</a><h4 class="code-header">fn <a href="#method.print" class="fnname">print</a>(&amp;mut self, _values: &amp;[<a class="enum" href="enum.Value.html" title="enum scad_runtime::Value">Value</a>]) -&gt; <a class="enum" href="https://doc.rust-lang.org/nightly/core/result/enum.Result.html" title="enum core::result::Result">Result</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.unit.html">()</a>, <a class="enum" href="enum.RuntimeError.html" title="enum scad_runtime::RuntimeError">RuntimeError</a>&gt;</h4></div></div><h2 id="foreign-impls" class="small-section-header">Implementations on Foreign Types<a href="#foreign-impls" class="anchor"></a></h2><details class="rustdoc-toggle implementors-toggle"><summary><section id="impl-Callbacks-for-%26mut%20C" class="impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#51">source</a><a href="#impl-Callbacks-for-%26mut%20C" class="anchor"></a><h3 class="code-header in-band">impl&lt;C:&nbsp;<a class="trait" href="trait.Callbacks.html" title="trait scad_runtime::Callbacks">Callbacks</a> + ?<a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a>&gt; <a class="trait" href="trait.Callbacks.html" title="trait scad_runtime::Callbacks">Callbacks</a> for <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.reference.html">&amp;mut </a>C</h3></section></summary><div class="impl-items"><section id="method.before_execute-1" class="method trait-impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#51">source</a><a href="#method.before_execute-1" class="anchor"></a><h4 class="code-header">fn <a href="#method.before_execute" class="fnname">before_execute</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;&amp;mut self,<br>&nbsp;&nbsp;&nbsp;&nbsp;current_chunk: &amp;<a class="struct" href="../scad_bytecode/struct.Chunk.html" title="struct scad_bytecode::Chunk">Chunk</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;instruction_pointer: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.usize.html">usize</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;instruction: <a class="enum" href="../scad_bytecode/enum.Instruction.html" title="enum scad_bytecode::Instruction">Instruction</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;stack: &amp;<a class="struct" href="struct.Stack.html" title="struct scad_runtime::Stack">Stack</a><br>)</h4></section><section id="method.consume_geometry-1" class="method trait-impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#51">source</a><a href="#method.consume_geometry-1" class="anchor"></a><h4 class="code-header">fn <a href="#method.consume_geometry" class="fnname">consume_geometry</a>(&amp;mut self, geometry: <a class="enum" href="enum.Geometry.html" title="enum scad_runtime::Geometry">Geometry</a>)</h4></section></div></details><details class="rustdoc-toggle implementors-toggle"><summary><section id="impl-Callbacks-for-Box%3CC%3E" class="impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#52">source</a><a href="#impl-Callbacks-for-Box%3CC%3E" class="anchor"></a><h3 class="code-header in-band">impl&lt;C:&nbsp;<a class="trait" href="trait.Callbacks.html" title="trait scad_runtime::Callbacks">Callbacks</a> + ?<a class="trait" href="https://doc.rust-lang.org/nightly/core/marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a>&gt; <a class="trait" href="trait.Callbacks.html" title="trait scad_runtime::Callbacks">Callbacks</a> for <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/boxed/struct.Box.html" title="struct alloc::boxed::Box">Box</a>&lt;C&gt;</h3></section></summary><div class="impl-items"><section id="method.before_execute-2" class="method trait-impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#52">source</a><a href="#method.before_execute-2" class="anchor"></a><h4 class="code-header">fn <a href="#method.before_execute" class="fnname">before_execute</a>(<br>&nbsp;&nbsp;&nbsp;&nbsp;&amp;mut self,<br>&nbsp;&nbsp;&nbsp;&nbsp;current_chunk: &amp;<a class="struct" href="../scad_bytecode/struct.Chunk.html" title="struct scad_bytecode::Chunk">Chunk</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;instruction_pointer: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.usize.html">usize</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;instruction: <a class="enum" href="../scad_bytecode/enum.Instruction.html" title="enum scad_bytecode::Instruction">Instruction</a>,<br>&nbsp;&nbsp;&nbsp;&nbsp;stack: &amp;<a class="struct" href="struct.Stack.html" title="struct scad_runtime::Stack">Stack</a><br>)</h4></section><section id="method.consume_geometry-2" class="method trait-impl has-srclink"><a class="srclink rightside" href="../src/scad_runtime/callbacks.rs.html#52">source</a><a href="#method.consume_geometry-2" class="anchor"></a><h4 class="code-header">fn <a href="#method.consume_geometry" class="fnname">consume_geometry</a>(&amp;mut self, geometry: <a class="enum" href="enum.Geometry.html" title="enum scad_runtime::Geometry">Geometry</a>)</h4></section></div></details><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor"></a></h2><div class="item-list" id="implementors-list"></div><script src="../implementors/scad_runtime/callbacks/trait.Callbacks.js" data-ignore-extern-crates="alloc" async></script></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="scad_runtime" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.65.0-nightly (b44197abb 2022-09-05)" ></div></body></html>