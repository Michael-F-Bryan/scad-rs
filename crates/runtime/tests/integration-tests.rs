use scad_bytecode::Program;
use scad_compiler::{lowering::Lowering, parsing::Parsing, Database};
use scad_runtime::{Geometry, RuntimeError, Value, VirtualMachine};

#[test]
fn run_pass() {
    insta::glob!("run-pass/*.scad", |path| {
        let src = std::fs::read_to_string(path).unwrap();
        let program = compile(&src);

        let mut output = Output::default();
        let mut vm = VirtualMachine::load(program);
        vm.run(Callbacks(&mut output)).expect("Execution failed");

        insta::assert_debug_snapshot!(output);
    });
}

#[test]
fn run_fail() {
    insta::glob!("run-fail/*.scad", |path| {
        let src = std::fs::read_to_string(path).unwrap();
        let program = compile(&src);

        let mut output = Output::default();
        let mut vm = VirtualMachine::load(program);
        let error = vm
            .run(Callbacks(&mut output))
            .expect_err("Execution should have failed");

        #[derive(Debug)]
        #[allow(dead_code)]
        struct Result {
            output: Output,
            error: RuntimeError,
        }

        insta::assert_debug_snapshot!(Result { output, error });
    });
}

fn compile(src: &str) -> Program {
    let mut db = Database::default();
    db.set_src(src.into());
    let (program, diags) = db.compile();

    if diags.has_warnings() {
        panic!("The program didn't compile cleanly: {diags:#?}");
    }

    program
}

struct Callbacks<'a>(&'a mut Output);

impl scad_runtime::Callbacks for Callbacks<'_> {
    fn consume_geometry(&mut self, geometry: Geometry) {
        self.0.geometry.push(geometry);
    }

    fn print(&mut self, values: &[Value]) -> Result<(), scad_runtime::RuntimeError> {
        self.0.printed.push(values.to_vec());
        Ok(())
    }
}

#[derive(Debug, Default, Clone)]
pub struct Output {
    geometry: Vec<Geometry>,
    printed: Vec<Vec<Value>>,
}
