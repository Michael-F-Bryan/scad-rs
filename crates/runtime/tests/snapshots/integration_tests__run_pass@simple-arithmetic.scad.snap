---
source: crates/runtime/tests/integration-tests.rs
expression: output
input_file: crates/runtime/tests/run-pass/simple-arithmetic.scad
---
Output {
    geometry: [],
    printed: [
        [
            String(
                "add",
            ),
            Number(
                2.0,
            ),
        ],
        [
            String(
                "sub",
            ),
            Number(
                4.0,
            ),
        ],
        [
            String(
                "mul",
            ),
            Number(
                4.5,
            ),
        ],
        [
            String(
                "div",
            ),
            Number(
                0.6666666666666666,
            ),
        ],
        [
            String(
                "neg",
            ),
            Number(
                -1.5,
            ),
        ],
        [
            String(
                "add mul",
            ),
            Number(
                12.0,
            ),
        ],
        [
            String(
                "mul add",
            ),
            Number(
                9.0,
            ),
        ],
        [
            String(
                "mul div",
            ),
            Number(
                0.8,
            ),
        ],
        [
            String(
                "one div zero",
            ),
            Number(
                inf,
            ),
        ],
        [
            String(
                "negative one div zero",
            ),
            Number(
                -inf,
            ),
        ],
        [
            String(
                "zero div zero",
            ),
            Number(
                NaN,
            ),
        ],
    ],
}
